crates/sui-core/src/consensus_adapter.rs

```diff
@@ -292,21 +293,19 @@ impl ConsensusAdapter {
                 .tap_err(|r| {
                     error!("Submit transaction failed with: {:?}", r);
                 })?;
-
+            self.num_inflight_transactions
+                .fetch_add(1, Ordering::SeqCst);
             // Increment the attempted certificate sequencing
             self.opt_metrics.as_ref().map(|metrics| {
                 metrics.sequencing_certificate_attempt.inc();
+                metrics
+                    .sequencing_certificate_inflight
+                    .set(self.num_inflight_transactions.load(Ordering::Relaxed) as i64);
             });
         }
 
-        // Wait for the consensus to sequence the certificate and assign locks to shared objects.
-        // Since the consensus protocol may drop some messages, it is not guaranteed that our
-        // certificate will be sequenced. So the best we can do is to set a timer and notify the
-        // client to retry if we timeout without hearing back from consensus (this module does not
-        // handle retries). The best timeout value depends on the consensus protocol.
-        let back_off_delay =
-            self.delay_step + Duration::from_millis(self.delay_ms.load(Ordering::Relaxed));
-        let result = match timeout(back_off_delay, waiter.wait_for_result()).await {
+        // TODO: make consensus guarantee delivery after submit_transaction() returns, and avoid the timeout below.
```

Comment:
Hhm based on my so far understanding of Narwhal I am not sure if we can actually guarantee delivery under the current semantics (even if we try to bend enough Narwhal's limits towards this). Maybe @gdanezis can provide more info around how plausible that would be. I can see Narwhal having a way to say "I am overloaded - I can't accept your transaction, come back later" - but how strong the delivery guarantees could be is another matter. Otherwise that TODO might not be that realistic.

Reply:
As we discussed today and from @gdanezis's message yesterday, there seems to be a plausible path to this: we can make Narwhal workers reply to Sui after a transaction is in a batch, and make sure the batch is included in a certificate later. Since Sui and workers may live on different machines and not share fate, the worker would need to persist the batch before replying to Sui. But I agree definitely there are a lot of nuances and potentially tricky correctness issues that we will need to investigate carefully.